### PR TITLE
Stop validating disabled questions

### LINKED ISSUE #1760: SDC | Bug: hidden field showing in validation required highlight
**Describe the bug**
Hidden fields are being shown in "Errors found" dialog on clicking submit button

**Component**
Core library, SDC library, reference app, or SDC gallery app

**To Reproduce**
1. Go to This Repo: https://github.com/parthfloyd/android-fhir & open questionnaire in demo app. On hitting submit it will show the hidden question in "errors found" dialog"

**Expected behavior**
It should not show the error

**Screenshots**
hidden-error-screenshot

### PR DESCRIPTION
**Pull request by jingtang10**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1760

**Description**

- Move the logic in `findEnableWhenQuestionnaireResponseItem` from the `QuestionnarieViewModel` class to `EnablementEvaluator`. This is a better logical separation since that is business logic specific to how enableWhen works
- Related to the previous refactoring, the `questionnaireResponseItemPreOrderList` and `questionnaireResponseItemParentMap` are both moved to `EnablementEvaluator` too.
- `EnablementEvaluator` is no longer an `object` since it is now specific to a QuestionnaireResponse in order to build the structures required for finding the questionnaire response item that determins the enablement status of another question.
- As a result, `QuestionnaireResponseValidator` is now able to use `EnablementEvaluator` in its validation.

**Alternative(s) considered**
NA

**Type**
Bug fix

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt [hunk 1]
```diff
@@ -41,19 +43,72 @@ import org.hl7.fhir.r4.model.QuestionnaireResponse
...
+        for (answerItem in answer.item) {
+          buildPreOrderList(answerItem)
+        }
+      }
+    }
+
+    for (item in questionnaireResponse.item) {
+      buildPreOrderList(item)
+    }
+  }
```
Comment (omarismail94) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt:90]: in QuestionnaireViewModel, we replaced this logic with the much cleaner:

```
private val QuestionnaireResponse.descendant:
  List<QuestionnaireResponse.QuestionnaireResponseItemComponent>
  get() = item.flatMap { it.descendant }

private val QuestionnaireResponse.QuestionnaireResponseItemComponent.descendant:
  List<QuestionnaireResponse.QuestionnaireResponseItemComponent>
  get() {
    return listOf(this) +
      this.item.flatMap { it.descendant } +
      this.answer.flatMap { it.item.flatMap { it.descendant } }
  }
```

If `flatMap` guarantees order, should we do the same here?

File: datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt [hunk 2]
```diff
@@ -102,43 +150,79 @@ internal object EnablementEvaluator {
...
+   * Returns whether the `enableWhen` constraint is satisfied.
+   *
+   * @param questionnaireResponseItemRetriever function that returns the
+   * [QuestionnaireResponse.Item] with the `linkId`, or null if there isn't one.
+   */
+  private fun evaluateEnableWhen(
+    enableWhen: Questionnaire.QuestionnaireItemEnableWhenComponent,
+    questionnaireResponseItem: QuestionnaireResponse.QuestionnaireResponseItemComponent,
+  ): Boolean {
+    val target =
```
Comment (omarismail94) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt:170]: can this value be more descriptive? how about `targetQuestionnaireResponseItem`

**Review by omarismail94 — COMMENTED** [review]
Awesome change!!! Just a few comments

File: datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt [hunk 3]
```diff
@@ -102,43 +150,79 @@ internal object EnablementEvaluator {
...
-      enableWhen.answerBooleanType.booleanValue()
-  } else {
-    // The `enableWhen` constraint evaluates to true if at least one answer has a value that
-    // satisfies the `enableWhen` operator and answer, with the exception of the `Exists` operator.
-    // See https://www.hl7.org/fhir/valueset-questionnaire-enable-operator.html.
-    questionnaireResponseItem?.answer?.any { enableWhen.predicate(it) } ?: false
+  /**
+   * Returns whether the `enableWhen` constraint is satisfied.
+   *
+   * @param questionnaireResponseItemRetriever function that returns the
```
[the remark this answers is not in the document]
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt:163]: removed it according to https://kotlinlang.org/docs/coding-conventions.html#documentation-comments which suggests avoiding @param if possible by incorporating it into the main block.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt [hunk 4]
```diff
@@ -41,19 +43,72 @@ import org.hl7.fhir.r4.model.QuestionnaireResponse
  *     ],
  * ```
  * specifies that the [Questionnaire.QuestionnaireItemComponent] should be enabled only if the
- * question with ID `vitaminKgiven` has been answered.
+ * question with linkId `vitaminKgiven` has been answered.
```
[the remark this answers is not in the document]
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt:46]: the example is right before this text. do you suggest we use different link ids?

**Comment by jingtang10** [comment]
@omarismail94 @aditya-07 please take another look thanks!

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 5]
```diff
@@ -761,67 +705,22 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
     }
 
   /** Gets a list of [QuestionnairePage]s for a paginated questionnaire. */
-  private fun getQuestionnairePages() =
+  private fun getQuestionnairePages(): List<QuestionnairePage>? {
+    val enablementEvaluator = EnablementEvaluator(questionnaireResponse)
     if (questionnaire.isPaginated) {
```
[the remark this answers is not in the document]
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:710]: done

File: datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt [hunk 6]
```diff
@@ -102,43 +124,78 @@ internal object EnablementEvaluator {
...
+   * Returns whether the `enableWhen` constraint is satisfied for the `questionnaireResponseItem`.
+   */
+  private fun evaluateEnableWhen(
+    enableWhen: Questionnaire.QuestionnaireItemEnableWhenComponent,
+    questionnaireResponseItem: QuestionnaireResponse.QuestionnaireResponseItemComponent,
+  ): Boolean {
+    val targetQuestionnaireResponseItem =
+      findEnableWhenQuestionnaireResponseItem(questionnaireResponseItem, enableWhen.question)
+    return if (Questionnaire.QuestionnaireItemOperator.EXISTS == enableWhen.operator) {
+      (targetQuestionnaireResponseItem == null ||
```
[the remark this answers is not in the document]
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt:144]: the thing is if the boolean value is false, that means the enable when needs to evaluate to true iff the answer doesn't exist. hence `targetQuestionnaireResponseItem.answer.isEmpty()) != enableWhen.answerBooleanType.booleanValue()`

i did switch the logic around a bit and added a comment - is this better?

**Review by stevenckngaa — APPROVED** [review]

**Pull request merged**